core/internal/runfiles/uploader.go

```diff
@@ -157,6 +182,87 @@ func (u *uploader) lockForOperation(method string) bool {
 	return true
 }
 
+// Begins watching the given path and uploading when the file changes.
+//
+// Note that this uses native filesystem events, and
+func (u *uploader) watch(path string) error {
+	// Lazily start the watcher when we receive our first file to watch.
+	if u.watcherOrNil == nil {
+		if err := u.startWatcher(); err != nil {
+			return err
+		}
+	}
+
+	if err := u.watcherOrNil.Add(path); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Starts up the file watcher goroutine.
+func (u *uploader) startWatcher() error {
+	if u.watcherOrNil != nil {
+		return fmt.Errorf(
+			"runfiles: tried to start watcher, but it is already started",
+		)
+	}
+
+	u.watcherOrNil = watcher.New()
+	u.watcherOrNil.FilterOps(watcher.Write)
+
+	killLoop := make(chan struct{})
+	u.watcherWG.Add(2)
+
+	go func() {
+		defer u.watcherWG.Done()
+
+		u.loopWatchFiles(killLoop)
```

Comment:
why is it better than using context?

Reply:
or maybe error group?

Reply:
oh interesting, i didn't know about errgroup! i'll try that instead